indexer/src/rmq_publisher.rs

```diff
@@ -95,11 +99,72 @@ impl RabbitPublisher {
         })
     }
 
-    pub fn run(&self, receiver: mpsc::Receiver<PublishData>) {
-        actix::spawn(self.clone().publisher(receiver));
+    pub fn run(&self, receiver: mpsc::Receiver<PublishData>) -> JoinHandle<()> {
+        let task = RmqPublisherTask::new(self.connection_pool.clone(), self.listener.clone(), receiver);
+
+        actix::spawn(task.run())
+    }
+}
+
+impl Metricable for RmqPublisher {
+    fn enable_metrics(&mut self, registry: Registry) -> Result<()> {
+        let listener = make_publisher_metrics(registry)?;
+        self.listener = Some(listener);
+
+        Ok(())
+    }
+}
+
+enum RmqPublisherState {
+    Shutdown,
+    WaitingForConnection,
+    Connected { connection: Connection },
+}
+
+struct RmqPublisherTask {
+    connection_pool: Pool,
+    receiver: mpsc::Receiver<PublishData>,
+    listener: Option<PublisherListener>,
+}
+
+impl RmqPublisherTask {
+    pub fn new(
+        connection_pool: Pool,
+        listener: Option<PublisherListener>,
+        receiver: mpsc::Receiver<PublishData>,
+    ) -> Self {
+        Self {
+            connection_pool,
+            receiver,
+            listener,
+        }
+    }
+
+    pub async fn run(mut self) {
```

Comment:
Yeah, this approach is preferred, to have a struct, an entrypoint method, and extra methods to handle particular cases.